Added option for parsing dates from filenames

taobrienlbl opened the pull request:
The new cfreader::filename_time_template option allows TECA to infer time information from filenames.  This is particularly for the ARTMIP MERRA2 data, which doesn't have a time coordinate. This runs like follows:

`teca_bayesian_ar_detect --input_regex "/project/projectdirs/m1517/cascade/external_datasets/ARTMIP/MERRA 2D/1983/ARTMIP_MERRA_2D_198309\.\*.nc" --cf_reader::filename_time_template "ARTMIP_MERRA_2D_%Y%m%d_%H.nc" --cf_reader::t_axis_variable ""`

burlen commented:
forgot to mention, rebase from master before merge, this will also very likely fix failing regression tests

taobrienlbl commented:
@burlen - I refactored the if-statement, fixed tab/whitespace issues, and squashed commits.  Tests are running now. Could you please review again and/or approve once tests are complete?  Thanks
